clippy_lints/src/needless_move.rs

```diff
@@ -0,0 +1,205 @@
+
+            captured.remove(hir_id);
+        }
+
+        let lint = |note_msg: &'static str| {
+            span_lint_and_then(
+                cx,
+                NEEDLESS_MOVE,
+                expr.span,
+                "you seem to use `move`, but the `move` is unnecessary",
+                |diag| {
+                    diag.suggest_remove_item(cx, move_kw, "remove the `move`", Applicability::MachineApplicable);
+                    diag.note(note_msg);
+                },
+            );
+        };
+
+        match (moved.is_empty(), captured.is_empty(), mutated.is_empty()) {
+            (true, true, true) => {
+                lint("there were no captured variables, so the `move` is unnecessary");
+            },
+            (false, true, _) => {
+                lint("there were consumed variables, but no borrowed variables, so the `move` is unnecessary");
+            },
+            (_, false, _) | (_, _, false) => {
+                // captured_vars is not empty, or mutated_vars is not empty, so `move` actually
+                // makes a difference and we should not remove it from this closure.
+            },
+        }
+    }
+}
+
+impl<'tcx> LateLintPass<'tcx> for NeedlessMove {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if expr.span.from_expansion() {
+            return;
+        }
+
+        let ExprKind::Closure(closure) = &expr.kind else {
+            return;
+        };
+
+        Self::check_closure(cx, expr, closure);
+    }
+}
+
+struct MovedVariablesCtxt {
+    // for each base variable, we remember:
+    /// The places where it was captured (and consumed, e.g. moved into the closure).
+    moved: FxIndexMap<HirId, Vec<ty::UpvarId>>,
+    /// The places where it was captured by reference (and not consumed).
+    /// If this vector is not empty, then the `move` keyword makes a difference.
+    /// There are a few special cases though, such as:
+    /// 1. We also handle the case in which there's both a borrow and a move of the
+    /// same value into the closure, e.g.:
+    ///
+    /// ```no_run
+    /// let x = String::new();
+    /// let closure = move || {
+    ///     let s = x.as_str(); // L1
+    ///     println!("{s}");
+    ///     drop(x); // L2
+    /// };
+    /// ```
+    ///
+    /// In this case, the `x` `String` gets moved into the closure (because of L2), but
+    /// it is also borrowed prior to that at L1.
+    ///
+    /// How we handle this is by removing the entries that point to `x` if it was captured
+    /// by value (and therefore moved into the closure).
+    captured: FxIndexMap<HirId, Vec<(ty::UpvarId, HirId, ty::BorrowKind)>>,
+
+    mutated: FxIndexMap<HirId, Vec<(ty::UpvarId, HirId)>>,
+}
+
+impl MovedVariablesCtxt {
+    fn move_common(&mut self, cmt: &euv::PlaceWithHirId<'_>, _: HirId) {
+        if let euv::PlaceBase::Upvar(vid) = cmt.place.base {
+            self.moved.entry(vid.var_path.hir_id).or_default().push(vid);
```

Comment:
Can you add this as a test? I suspect that this is currently a FP but I could be wrong.
```rs
let v = (String::new(), String::new());
assert_static(move || {
  let _w = v.0;
  let _h = &v.1;
});
```

Reply:
Yeah, that's indeed a FP; gonna look into it

Reply:
This should be fixed now